File: bot/strategies/urlPatterns.js
import fetch from 'node-fetch';
import * as cheerio from 'cheerio';

const PATTERNS = [
  '/coupons', '/promo', '/promo-codes', '/discount', '/deals',
  '/offers', '/special-offers', '/promotions', '/sale',
  '/coupon-codes', '/voucher', '/vouchers', '/promo-code',
  '/student-discount', '/referral', '/referral-program',
  '/black-friday', '/cyber-monday', '/holiday-sale',
  '/pricing', '/plans', '/subscription',
];

const KEYWORDS = [
  'coupon', 'promo code', 'promo', 'discount', 'deal', 'offer',
  'save', 'sale', 'voucher', 'referral', 'refer', 'bonus',
  'free trial', 'free shipping', 'student', 'first order',
];

// Regex patterns for actual coupon codes
const CODE_PATTERNS = [
  /\b[A-Z0-9]{4,20}\b/g,              // SAVE20, WELCOME10, FREESHIP
  /(?:code|coupon|promo)[:\s]+([A-Z0-9]{4,20})/gi,
  /(?:use|enter|apply)\s+code[:\s]+([A-Z0-9]{4,20})/gi,
];

// Discount extraction patterns
const DISCOUNT_PATTERNS = [
  /(\d+%)\s*off/i,
  /\$(\d+)\s*off/i,
  /save\s+\$?(\d+|up\s+to\s+\d+%)/i,
  /free\s+(trial|shipping|delivery)/i,
  /up\s+to\s+(\d+%)/i,
  /(\d+%)\s+discount/i,
];

const UAs = [
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120.0.0.0',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 Chrome/120.0.0.0',
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 Chrome/120.0.0.0',
];

function ua() { return UAs[Math.floor(Math.random() * UAs.length)]; }

function extractCodes(text) {
  const found = new Set();
  for (const pattern of CODE_PATTERNS) {
    const matches = text.matchAll(pattern);
    for (const m of matches) {
      const code = (m[1] || m[0]).trim();
      if (code.length >= 4 && code.length <= 20 && /[A-Z0-9]{4,}/i.test(code)) {
        found.add(code.toUpperCase());
      }
    }
  }
  return [...found];
}

function extractDiscount(text) {
  for (const pattern of DISCOUNT_PATTERNS) {
    const m = text.match(pattern);
    if (m) return m[0];
  }
  return null;
}

function extractTitle($) {
  const h1 = $('h1').first().text().trim();
  if (h1 && h1.length > 5 && h1.length < 200) return h1;
  const title = $('title').text().trim();
  if (title) return title;
  return null;
}

export async function discoverByUrlPatterns(brand) {
  const { brandName, website } = brand;
  const base = website.replace(/\/$/, '');
  const results = [];

  for (const pattern of PATTERNS) {
    const url = `${base}${pattern}`;
    try {
      const res = await fetch(url, {
        headers: { 'User-Agent': ua() },
        redirect: 'follow',
        follow: 3,
        timeout: 8000,
      });
      if (!res.ok) continue;

      const html = await res.text();
      const $ = cheerio.load(html);
      const text = $('body').text().toLowerCase();

      const matched = KEYWORDS.filter(kw => text.includes(kw));
      if (matched.length < 2 && !text.match(/[A-Z0-9]{4,20}/)) continue;

      // Check normalized URL for dedup
      const norm = url.replace(/\/$/, '').toLowerCase();
      if (results.some(r => r.sourceUrl.replace(/\/$/, '').toLowerCase() === norm)) continue;

      const title = extractTitle($) || `${brandName} ${pattern}`;
      const codes = extractCodes(text);
      const discount = extractDiscount(text);

      // Boost confidence if we found actual codes
      let confidence = Math.min(50 + matched.length * 8, 98);
      if (codes.length > 0) confidence = Math.min(confidence + 15, 99);
      if (discount) confidence = Math.min(confidence + 5, 99);

      results.push({
        sourceUrl: url,
        sourcePage: pattern,
        sourceReliability: codes.length > 0 ? 'Official Site' : 'Official Site',
        confidence,
        title: title.slice(0, 200),
        description: ($('meta[name="description"]').attr('content') || '').slice(0, 500),
        discount: discount || 'Special offer',
        codes: [...new Set(codes)],
      });
    } catch {
      // Timeout or network error — skip silently
    }
  }

  return results;
}

File: bot/app/adapters/url_patterns.py
import asyncio, re
from bs4 import BeautifulSoup
from app.extractors import extract_codes_from_soup, detect_countries
from app.services.fetcher import fetch_direct, is_cloudflare

PATTERNS = [
    '/coupons','/promo','/promo-codes','/promo-code','/discount',
    '/deals','/offers','/promotions','/sale',
    '/coupon-codes','/coupon','/voucher','/vouchers',
    '/new-customer','/welcome','/first-order',
    '/signup-offer','/sign-up','/free-trial','/freetrial',
    '/student','/student-discount','/students',
    '/education','/referral','/refer',
    '/rewards','/loyalty','/cashback','/gift-card',
    '/pricing','/plans','/subscription',
    '/black-friday','/flash-sale','/clearance',
    '/affiliate','/partner',
    '/free','/freebies','/free-shipping',
    '/bundle','/promos',
]
KEYWORDS = ['coupon','promo code','promo','discount','deal','offer','save','sale','voucher','free trial','student']
DISCOUNT_RE = re.compile(r'(\d+%)\s*off|\$\d+\s*off|free\s+(trial|shipping|delivery)', re.I)

class UrlPatternsAdapter:
    name = "urlPatterns"

    async def discover(self, brand: dict, client):
        base = brand['website'].rstrip('/')
        brand_name = brand['brandName']
        sem = asyncio.Semaphore(20)

        async def fetch(pattern):
            url = f"{base}{pattern}"
            async with sem:
                try:
                    result = await fetch_direct(client, url, timeout=4.0)
                    if result["status"] != 200:
                        return None if not result.get("blocked") else {
                            "sourceUrl": url, "blocked": True, "blocked_reason": "cloudflare",
                            "codes": [], "confidence": 0, "title": "", "discount": "", "countries": []
                        }
                    if is_cloudflare(result["text"]):
                        return {"sourceUrl": url, "blocked": True, "blocked_reason": "cloudflare",
                                "codes": [], "confidence": 0, "title": "", "discount": "", "countries": []}
                    soup = BeautifulSoup(result["text"], 'lxml')
                    text = soup.get_text().lower()
                    codes = extract_codes_from_soup(soup, url, brand_name)
                    kw = sum(1 for k in KEYWORDS if k in text)
                    if kw < 2 and not codes: return None
                    conf = 40 + kw * 6 + (20 if codes else 0)
                    title_el = soup.find('title')
                    title = title_el.get_text(strip=True) if title_el else ''
                    if any(k in title.lower() for k in KEYWORDS): conf += 10
                    conf = min(conf, 99)
                    dm = DISCOUNT_RE.search(text)
                    countries = detect_countries(url, result["text"], brand_name)
                    return {
                        "sourceUrl": url, "sourcePage": pattern, "countries": countries,
                        "confidence": conf, "title": title[:200], "description": "",
                        "discount": dm.group(0) if dm else "Special offer",
                        "codes": codes, "blocked": False,
                    }
                except:
                    return None

        results = await asyncio.gather(*[fetch(p) for p in PATTERNS])
        return [r for r in results if r]
